main.cpp

```diff
@@ -241,39 +119,73 @@ struct HeapA
  Wait for my code review.
  */
 
-template<typename T>
+#include <iostream>
+#include <functional>
+#include <memory>
+#include <cmath>
+#include <typeinfo>
+#include <type_traits>
+
+template<typename NumericType>
+struct Temporary
+{
+    Temporary(NumericType t) : v(t)
+    {
+        std::cout << "I'm a Temporary<" << typeid(v).name() << "> object, #"
+                  << counter++ << std::endl;
+    }
+    operator NumericType&() const { return v;}
+    operator NumericType&() { return v; }
+private:
+    static int counter;
+    NumericType v;
+};
+
+template<typename NumericType>
+int Temporary<NumericType>::counter = 0;
+
+
+template<typename FirstType>
```

Comment:
You didn't need to change the template parameter's name
Don't revise code that doesn't need revising
- [ ] revert